pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,326 @@
+					resp.StatusCode))
+				continue
+			}
+			if err := validateProtoMessage(el.actualURL, resp); err != nil {
+				// send something to ledmanager ???
+				errList = append(errList, fmt.Sprintf("validateProtoMessage: resp header error: %s", err))
+				continue
+			}
+			localProfile, err := parseAndValidateLocalProfile(contents, getconfigCtx)
+			if err != nil {
+				errList = append(errList, fmt.Sprintf("parseAndValidateLocalProfile: %s", err))
+				continue
+			}
+			return localProfile, nil
+		}
+	}
+	return nil, fmt.Errorf("getLocalProfileConfig: all attempts failed: %s", strings.Join(errList, ";"))
+}
+
+func writeReceivedProtoMessageLocalProfile(localProfile *profile.LocalProfile) {
+	contents, err := proto.Marshal(localProfile)
+	if err != nil {
+		log.Errorf("writeReceivedProtoMessageLocalProfile Marshalling failed: %s", err)
+		return
+	}
+	writeProtoMessage(savedLocalProfileFile, contents)
+	return
+}
+
+//parseProfile process local and global profile configuration
+//must be called before processing of app instances from config
+func parseProfile(ctx *getconfigContext, config *zconfig.EdgeDevConfig) {
+	log.Functionf("parseProfile start: globalProfile: %s localProfile: %s",
+		ctx.globalProfile, ctx.localProfile)
+	if ctx.globalProfile != config.GlobalProfile {
+		log.Noticef("parseProfile: GlobalProfile changed from %s to %s",
+			ctx.globalProfile, config.GlobalProfile)
+		ctx.globalProfile = config.GlobalProfile
+	}
+	if ctx.localProfileServer != config.LocalProfileServer {
+		// read saved config on the first call
+		if !ctx.callProcessProfile {
+			if localProfile, err := readSavedLocalProfile(ctx, false); err != nil {
+				log.Functionf("parseProfile readSavedLocalProfile: %s", err)
+			} else if localProfile != nil {
+				log.Noticef("starting with localProfile %s", localProfile.LocalProfile)
+				ctx.localProfile = localProfile.LocalProfile
+			}
+			ctx.callProcessProfile = true
+		} else {
+			// if localProfileServer changed not in the first call of function
+			// clean saved proto message and set localProfile to empty
+			cleanSavedProtoMessage(savedLocalProfileFile)
+			ctx.localProfile = ""
+		}
+		log.Noticef("parseProfile: LocalProfileServer changed from %s to %s",
+			ctx.localProfileServer, config.LocalProfileServer)
+		ctx.localProfileServer = config.LocalProfileServer
+		if ctx.localProfileServer != "" {
+			// notify to reload local profile on profileServerChange
+			select {
+			case ctx.localProfileTrigger <- Notify{}:
+			default:
+			}
+		}
+	}
+	ctx.profileServerToken = config.ProfileServerToken
+	currentProfile := determineCurrentProfile(ctx)
+	if ctx.currentProfile != currentProfile {
+		// if we do not use LocalProfileServer set profile to GlobalProfile
+		log.Noticef("parseProfile: profile changed from %s to %s",
+			ctx.currentProfile, currentProfile)
+		ctx.currentProfile = currentProfile
+		publishZedAgentStatus(ctx)
+	}
```

Comment:
No, this belong as the top and not in a parse function. I thought we agreed on that earlier today.

Reply:
In this case we will publish appInstanceConfigs before (concurrently) than current context, so, zedmanager will activate apps based on Activate (will start additional apps, we may not want to start based on currentProfile).

Reply:
Which case are you concerned about? The globalProfile changing in the same EdgeDevConfig as some set of app instances are added or have their profile changed?

The system is asynchronous and with eventual consistency, so there might be some uncommon cases when there will be some churn, but it will converge.
What matters is that in a normal operation (such as a device reboot) there isn't any churn, which is why it is crtical that the code which reads the saved profile is well-understood to run before anything else.

Reply:
Multiple changes in the same update to EdgeDevConfig can result in some churn, but the system will be eventually consistent. Thus only question is in which cases we must avoid the churn (which I think is mostly about the reboot of the device).

Reply:
I concern about the case after reboot: we have empty local profile, and we do publish of app instance configs having current profile based on global only. So, we must read saved one here and publish zedagentstatus to zedmanager if local profile server defined. We should have local profile here to prepare correct currentProfile.

Reply:
This is why we should have the top of the task read the current profile. (Waiting for the trigger was a bad idea to handle that case.) Up front it can read it and set ctx.localProfile, ctx.currentProfile.

The trigger is useful to have the task react to changes (in localProfileServer or globalProfile) more quickly.